# Task 1: Data Cleaning and Preprocessing

## Hospital Patient Records

###  improt the required library

In [1]:
# import numpy 
# this library help us to perform numerical operation on diffrent data structure

import numpy as np

# import pandas 
# this library help us to convert diffrent data entry in pandas data frame 

import pandas as pd
# import matplotlib and seaborn
# this library help us visulization diffrent data structure

import matplotlib.pyplot as plt
import seaborn as sns

### Load The  Raw Dataset

In [2]:
# read the data from the local 

df_raw_dataset_Healthcare_Patient_Records =pd.read_csv('raw_dataset_Healthcare_Patient_Records.csv')

In [3]:
# show the first five file observations of given data frame 
df_raw_dataset_Healthcare_Patient_Records.head()

,Patient_ID,Patient_Name,Age,Gender,Blood_Group,City,Department,Diagnosis,Admission_Date,Discharge_Date,...,BMI,Smoking_Status,Diabetes,Hypertension,Medication,Treatment_Cost_INR,Insurance_Type,Doctor_ID,Outcome,Admission_Type
0,P106252,Sneha Singh,31.0,Male,AB+,Ahmedabad,Oncology,Gastritis,2025-04-18,2025-04-20,...,15.7,Never,Yes,No,Insulin,13238.96,Government,D696,Recovered,Elective
1,P104684,Vivaan Reddy,38.0,Male,B-,Kolkata,Orthopedics,Arthritis,2022-05-09,2022-05-12,...,34.5,Never,No,No,Metformin,8232.69,Employer,D738,Improved,Emergency
2,P101731,Isha Mehta,67.0,Male,AB-,Rajkot,Neurology,Diabetes,2024-09-12,2024-09-23,...,29.1,Never,No,No,Amoxicillin,23295.42,Government,D350,Referred,Emergency
3,P104742,Rahul Koli,53.0,M,O+,Ahmedabad,Orthopedics,Gastritis,2023-04-07,2023-04-13,...,23.3,Never,No,No,Metformin,25632.82,Private,D634,Referred,Urgent
4,P104521,Vikram Desai,39.0,Male,B+,Vadodara,Oncology,Diabetes,2022-10-05,2022-10-08,...,35.5,Current,No,Yes,Paracetamol,20111.41,Government,D737,Recovered,Elective


In [4]:
# show the last five file observations of given data frame 
df_raw_dataset_Healthcare_Patient_Records.tail()

,Patient_ID,Patient_Name,Age,Gender,Blood_Group,City,Department,Diagnosis,Admission_Date,Discharge_Date,...,BMI,Smoking_Status,Diabetes,Hypertension,Medication,Treatment_Cost_INR,Insurance_Type,Doctor_ID,Outcome,Admission_Type
9995,P105734,Vikram Reddy,37.0,Male,A+,Rajkot,Oncology,Fracture,2023-12-22,2024-01-03,...,29.5,Never,Yes,No,Metformin,12065.84,Self-Pay,D415,Recovered,ELECTIVE
9996,P105191,Pooja Singh,56.0,Female,AB-,Vadodara,Pediatrics,Hypertension,2025-10-24,2025-10-27,...,16.1,Never,No,No,Atorvastatin,15477.88,Private,D781,Recovered,Elective
9997,P105390,Rahul Koli,31.0,Female,B+,Pune,Pediatrics,Fracture,2022-03-12,2022-03-23,...,32.4,Never,Yes,Yes,Amoxicillin,59541.18,Employer,D131,Recovered,emergency
9998,P100860,Neha Patel,49.0,Female,B+,Hyderabad,Dermatology,Migraine,2023-01-09,2023-01-18,...,15.7,Former,No,No,Amlodipine,30732.58,Employer,D155,Recovered,Elective
9999,P107270,Isha Reddy,23.0,Male,AB-,Hyderabad,ENT,Fracture,2022-05-17,2022-05-30,...,28.9,Current,No,No,Insulin,23623.18,Government,D767,Recovered,Emergency


### Data Overview

In [5]:
# check the shap of the data
df_raw_dataset_Healthcare_Patient_Records.shape

(10000, 30)

**Interpretation**
- we have 10000 observations and 30 attributes.

In [6]:
# check the columns present in the data
df_raw_dataset_Healthcare_Patient_Records.columns

Index(['Patient_ID', 'Patient_Name', 'Age', 'Gender', 'Blood_Group', 'City',
       'Department', 'Diagnosis', 'Admission_Date', 'Discharge_Date',
       'Length_of_Stay', 'Heart_Rate', 'Systolic_BP', 'Diastolic_BP',
       'Temperature_C', 'SpO2', 'Respiratory_Rate', 'Glucose_mg_dL',
       'Cholesterol_mg_dL', 'Hemoglobin_g_dL', 'BMI', 'Smoking_Status',
       'Diabetes', 'Hypertension', 'Medication', 'Treatment_Cost_INR',
       'Insurance_Type', 'Doctor_ID', 'Outcome', 'Admission_Type'],
      dtype='object')

In [7]:
# check the size and data type of the columns 
df_raw_dataset_Healthcare_Patient_Records.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 30 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Patient_ID          10000 non-null  object 
 1   Patient_Name        10000 non-null  object 
 2   Age                 9800 non-null   float64
 3   Gender              9808 non-null   object 
 4   Blood_Group         9801 non-null   object 
 5   City                9801 non-null   object 
 6   Department          9800 non-null   object 
 7   Diagnosis           9798 non-null   object 
 8   Admission_Date      10000 non-null  object 
 9   Discharge_Date      10000 non-null  object 
 10  Length_of_Stay      10000 non-null  int64  
 11  Heart_Rate          9799 non-null   float64
 12  Systolic_BP         10000 non-null  int64  
 13  Diastolic_BP        10000 non-null  int64  
 14  Temperature_C       10000 non-null  float64
 15  SpO2                10000 non-null  int64  
 16  Respi

**Interpretation** 
- we have 12 numerical columns and 18 categorical column
- the data size is 2.3+ MB
- we have missing data present in data frame

### Convert date formats to a consistent type (e.g., dd-mm-yyyy).

In [8]:
# To see the data in Admission Date and Discharge Date
df_raw_dataset_Healthcare_Patient_Records[['Admission_Date','Discharge_Date']].head()

,Admission_Date,Discharge_Date
0,2025-04-18,2025-04-20
1,2022-05-09,2022-05-12
2,2024-09-12,2024-09-23
3,2023-04-07,2023-04-13
4,2022-10-05,2022-10-08


In [9]:
# Convert the column to datetime
df_raw_dataset_Healthcare_Patient_Records['Admission_Date'] = pd.to_datetime(df_raw_dataset_Healthcare_Patient_Records['Admission_Date'],errors='coerce')

In [10]:
# Create separate Year, Month, Day columns
df_raw_dataset_Healthcare_Patient_Records['Admission_Year'] = (df_raw_dataset_Healthcare_Patient_Records['Admission_Date'].dt.year)

df_raw_dataset_Healthcare_Patient_Records['Admission_Month'] = (df_raw_dataset_Healthcare_Patient_Records['Admission_Date'].dt.month)

df_raw_dataset_Healthcare_Patient_Records['Admission_Day'] = (df_raw_dataset_Healthcare_Patient_Records['Admission_Date'].dt.day)

In [11]:
# Do the same for Discharge Date 
df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'] = pd.to_datetime(df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'],errors='coerce')

df_raw_dataset_Healthcare_Patient_Records['Discharge_Year'] = (df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'].dt.year)

df_raw_dataset_Healthcare_Patient_Records['Discharge_Month'] = (df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'].dt.month)

df_raw_dataset_Healthcare_Patient_Records['Discharge_Day'] = (df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'].dt.day)

In [12]:
# Chaek the result
df_raw_dataset_Healthcare_Patient_Records[
['Admission_Date','Admission_Year','Admission_Month','Admission_Day','Discharge_Date','Discharge_Year','Discharge_Month','Discharge_Day']].head()

,Admission_Date,Admission_Year,Admission_Month,Admission_Day,Discharge_Date,Discharge_Year,Discharge_Month,Discharge_Day
0,2025-04-18,2025,4,18,2025-04-20,2025,4,20
1,2022-05-09,2022,5,9,2022-05-12,2022,5,12
2,2024-09-12,2024,9,12,2024-09-23,2024,9,23
3,2023-04-07,2023,4,7,2023-04-13,2023,4,13
4,2022-10-05,2022,10,5,2022-10-08,2022,10,8


In [13]:
# if i want the day name, like Monday, Tuesday, etc. so
# Use .dt.day_name()
## for Admission Date
df_raw_dataset_Healthcare_Patient_Records['Admission_Date'] = pd.to_datetime(df_raw_dataset_Healthcare_Patient_Records['Admission_Date'],errors='coerce')

df_raw_dataset_Healthcare_Patient_Records['Admission_Day_Name'] = (df_raw_dataset_Healthcare_Patient_Records['Admission_Date'].dt.day_name())
## for  Discharge Date
df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'] = pd.to_datetime(df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'],errors='coerce')

df_raw_dataset_Healthcare_Patient_Records['Discharge_Day_Name'] = (df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'].dt.day_name())

In [14]:
# chek the result
df_raw_dataset_Healthcare_Patient_Records[['Admission_Date', 'Admission_Day_Name','Discharge_Date', 'Discharge_Day_Name']].head()

,Admission_Date,Admission_Day_Name,Discharge_Date,Discharge_Day_Name
0,2025-04-18,Friday,2025-04-20,Sunday
1,2022-05-09,Monday,2022-05-12,Thursday
2,2024-09-12,Thursday,2024-09-23,Monday
3,2023-04-07,Friday,2023-04-13,Thursday
4,2022-10-05,Wednesday,2022-10-08,Saturday


### Check and fix data types (e.g., age should be int, date as datetime)

In [15]:
# check current data type
# we also check with  info()
df_raw_dataset_Healthcare_Patient_Records.dtypes

Patient_ID                    object
Patient_Name                  object
Age                          float64
Gender                        object
Blood_Group                   object
City                          object
Department                    object
Diagnosis                     object
Admission_Date        datetime64[ns]
Discharge_Date        datetime64[ns]
Length_of_Stay                 int64
Heart_Rate                   float64
Systolic_BP                    int64
Diastolic_BP                   int64
Temperature_C                float64
SpO2                           int64
Respiratory_Rate               int64
Glucose_mg_dL                float64
Cholesterol_mg_dL              int64
Hemoglobin_g_dL              float64
BMI                          float64
Smoking_Status                object
Diabetes                      object
Hypertension                  object
Medication                    object
Treatment_Cost_INR            object
Insurance_Type                object
D

In [16]:
# Fix the important data types
# Age → numeric
df_raw_dataset_Healthcare_Patient_Records['Age'] = pd.to_numeric(df_raw_dataset_Healthcare_Patient_Records['Age'],errors='coerce').astype('Int64')
# Dates → datetime
df_raw_dataset_Healthcare_Patient_Records['Admission_Date'] = pd.to_datetime(df_raw_dataset_Healthcare_Patient_Records['Admission_Date'],errors='coerce')

df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'] = pd.to_datetime(
    df_raw_dataset_Healthcare_Patient_Records['Discharge_Date'],
    errors='coerce'
)



In [17]:
# check the result
df_raw_dataset_Healthcare_Patient_Records.dtypes

Patient_ID                    object
Patient_Name                  object
Age                            Int64
Gender                        object
Blood_Group                   object
City                          object
Department                    object
Diagnosis                     object
Admission_Date        datetime64[ns]
Discharge_Date        datetime64[ns]
Length_of_Stay                 int64
Heart_Rate                   float64
Systolic_BP                    int64
Diastolic_BP                   int64
Temperature_C                float64
SpO2                           int64
Respiratory_Rate               int64
Glucose_mg_dL                float64
Cholesterol_mg_dL              int64
Hemoglobin_g_dL              float64
BMI                          float64
Smoking_Status                object
Diabetes                      object
Hypertension                  object
Medication                    object
Treatment_Cost_INR            object
Insurance_Type                object
D

In [18]:
# Length of Stay → numeric
df_raw_dataset_Healthcare_Patient_Records['Length_of_Stay'] = pd.to_numeric(
    df_raw_dataset_Healthcare_Patient_Records['Length_of_Stay'],
    errors='coerce'
).astype('Int64')


# Vital signs / medical measurements → numeric
numeric_columns = [
    'Heart_Rate',
    'Systolic_BP',
    'Diastolic_BP',
    'Temperature_C',
    'SpO2',
    'Respiratory_Rate',
    'Glucose_mg_dL',
    'Cholesterol_mg_dL'
]

for col in numeric_columns:
    df_raw_dataset_Healthcare_Patient_Records[col] = pd.to_numeric(
        df_raw_dataset_Healthcare_Patient_Records[col],
        errors='coerce'
    )

In [19]:
# check the result
df_raw_dataset_Healthcare_Patient_Records.dtypes

Patient_ID                    object
Patient_Name                  object
Age                            Int64
Gender                        object
Blood_Group                   object
City                          object
Department                    object
Diagnosis                     object
Admission_Date        datetime64[ns]
Discharge_Date        datetime64[ns]
Length_of_Stay                 Int64
Heart_Rate                   float64
Systolic_BP                    int64
Diastolic_BP                   int64
Temperature_C                float64
SpO2                           int64
Respiratory_Rate               int64
Glucose_mg_dL                float64
Cholesterol_mg_dL              int64
Hemoglobin_g_dL              float64
BMI                          float64
Smoking_Status                object
Diabetes                      object
Hypertension                  object
Medication                    object
Treatment_Cost_INR            object
Insurance_Type                object
D

### Rename column headers to be clean and uniform (e.g., lowercase, no spaces).

In [20]:
# check the columns
df_raw_dataset_Healthcare_Patient_Records.columns

Index(['Patient_ID', 'Patient_Name', 'Age', 'Gender', 'Blood_Group', 'City',
       'Department', 'Diagnosis', 'Admission_Date', 'Discharge_Date',
       'Length_of_Stay', 'Heart_Rate', 'Systolic_BP', 'Diastolic_BP',
       'Temperature_C', 'SpO2', 'Respiratory_Rate', 'Glucose_mg_dL',
       'Cholesterol_mg_dL', 'Hemoglobin_g_dL', 'BMI', 'Smoking_Status',
       'Diabetes', 'Hypertension', 'Medication', 'Treatment_Cost_INR',
       'Insurance_Type', 'Doctor_ID', 'Outcome', 'Admission_Type',
       'Admission_Year', 'Admission_Month', 'Admission_Day', 'Discharge_Year',
       'Discharge_Month', 'Discharge_Day', 'Admission_Day_Name',
       'Discharge_Day_Name'],
      dtype='object')

In [21]:
# renaming or repalce in columns 
df_raw_dataset_Healthcare_Patient_Records.columns = (
    df_raw_dataset_Healthcare_Patient_Records.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace(r'[^a-z0-9_]', '', regex=True)
)

print(df_raw_dataset_Healthcare_Patient_Records.columns)

Index(['patient_id', 'patient_name', 'age', 'gender', 'blood_group', 'city',
       'department', 'diagnosis', 'admission_date', 'discharge_date',
       'length_of_stay', 'heart_rate', 'systolic_bp', 'diastolic_bp',
       'temperature_c', 'spo2', 'respiratory_rate', 'glucose_mg_dl',
       'cholesterol_mg_dl', 'hemoglobin_g_dl', 'bmi', 'smoking_status',
       'diabetes', 'hypertension', 'medication', 'treatment_cost_inr',
       'insurance_type', 'doctor_id', 'outcome', 'admission_type',
       'admission_year', 'admission_month', 'admission_day', 'discharge_year',
       'discharge_month', 'discharge_day', 'admission_day_name',
       'discharge_day_name'],
      dtype='object')


In [22]:
# check the first five record
df_raw_dataset_Healthcare_Patient_Records.head()

,patient_id,patient_name,age,gender,blood_group,city,department,diagnosis,admission_date,discharge_date,...,outcome,admission_type,admission_year,admission_month,admission_day,discharge_year,discharge_month,discharge_day,admission_day_name,discharge_day_name
0,P106252,Sneha Singh,31,Male,AB+,Ahmedabad,Oncology,Gastritis,2025-04-18,2025-04-20,...,Recovered,Elective,2025,4,18,2025,4,20,Friday,Sunday
1,P104684,Vivaan Reddy,38,Male,B-,Kolkata,Orthopedics,Arthritis,2022-05-09,2022-05-12,...,Improved,Emergency,2022,5,9,2022,5,12,Monday,Thursday
2,P101731,Isha Mehta,67,Male,AB-,Rajkot,Neurology,Diabetes,2024-09-12,2024-09-23,...,Referred,Emergency,2024,9,12,2024,9,23,Thursday,Monday
3,P104742,Rahul Koli,53,M,O+,Ahmedabad,Orthopedics,Gastritis,2023-04-07,2023-04-13,...,Referred,Urgent,2023,4,7,2023,4,13,Friday,Thursday
4,P104521,Vikram Desai,39,Male,B+,Vadodara,Oncology,Diabetes,2022-10-05,2022-10-08,...,Recovered,Elective,2022,10,5,2022,10,8,Wednesday,Saturday


### Standardize text values like gender, country names, etc.

In [23]:
# First check the unique text values
text_columns = [
    'gender',
    'blood_group',
    'city',
    'department',
    'diagnosis',
    'smoking_status',
    'diabetes',
    'hypertension',
    'medication',
    'insurance_type',
    'outcome',
    'admission_type'
]

for col in text_columns:
    print(f"\n{col}:")
    print(df_raw_dataset_Healthcare_Patient_Records[col].unique())


gender:
['Male' 'M' 'Female' nan 'Other' 'F' 'female']

blood_group:
['AB+' 'B-' 'AB-' 'O+' 'B+' 'O-' nan 'A-' 'A+']

city:
['Ahmedabad' 'Kolkata' 'Rajkot' 'Vadodara' 'Bengaluru' 'Hyderabad' 'Delhi'
 'Chennai' 'vadodara' 'Pune' 'Mumbai' 'Jaipur' 'Surat' nan 'Vadodara '
 'AHMEDABAD']

department:
['Oncology' 'Orthopedics' 'Neurology' 'Cardiology' 'Emergency'
 'Pediatrics' 'ENT' 'General Medicine' nan 'Dermatology']

diagnosis:
['Gastritis' 'Arthritis' 'Diabetes' 'Viral Infection' 'Heart Disease'
 'Pneumonia' 'Fracture' 'Hypertension' 'Asthma' 'Migraine' nan]

smoking_status:
['Never' 'Current' 'Former' 'current' nan]

diabetes:
['Yes' 'No']

hypertension:
['No' 'Yes']

medication:
['Insulin' 'Metformin' 'Amoxicillin' 'Paracetamol' nan 'Atorvastatin'
 'Amlodipine']

insurance_type:
['Government' 'Employer' 'Private' 'Self-Pay' nan]

outcome:
['Recovered' 'Improved' 'Referred' 'Critical']

admission_type:
['Elective' 'Emergency' 'Urgent' 'emergency' 'ELECTIVE']


In [24]:
# General text cleaning
# i want only  change in general
for col in text_columns:
    df_raw_dataset_Healthcare_Patient_Records[col] = (
        df_raw_dataset_Healthcare_Patient_Records[col]
        .astype('string')
        .str.strip()
        .str.lower()
    )
# Standardize Gender

    df_raw_dataset_Healthcare_Patient_Records['gender'] = (
    df_raw_dataset_Healthcare_Patient_Records['gender']
    .replace({
        'm': 'male',
        'man': 'male',
        'male': 'male',
        'f': 'female',
        'woman': 'female',
        'female': 'female'
    })
)

In [25]:
# check the result 
df_raw_dataset_Healthcare_Patient_Records['gender'].value_counts(dropna=False)

gender
male      4807
female    4797
other      204
<NA>       192
Name: count, dtype: Int64

In [26]:
# check the first five record
df_raw_dataset_Healthcare_Patient_Records.head()

,patient_id,patient_name,age,gender,blood_group,city,department,diagnosis,admission_date,discharge_date,...,outcome,admission_type,admission_year,admission_month,admission_day,discharge_year,discharge_month,discharge_day,admission_day_name,discharge_day_name
0,P106252,Sneha Singh,31,male,ab+,ahmedabad,oncology,gastritis,2025-04-18,2025-04-20,...,recovered,elective,2025,4,18,2025,4,20,Friday,Sunday
1,P104684,Vivaan Reddy,38,male,b-,kolkata,orthopedics,arthritis,2022-05-09,2022-05-12,...,improved,emergency,2022,5,9,2022,5,12,Monday,Thursday
2,P101731,Isha Mehta,67,male,ab-,rajkot,neurology,diabetes,2024-09-12,2024-09-23,...,referred,emergency,2024,9,12,2024,9,23,Thursday,Monday
3,P104742,Rahul Koli,53,male,o+,ahmedabad,orthopedics,gastritis,2023-04-07,2023-04-13,...,referred,urgent,2023,4,7,2023,4,13,Friday,Thursday
4,P104521,Vikram Desai,39,male,b+,vadodara,oncology,diabetes,2022-10-05,2022-10-08,...,recovered,elective,2022,10,5,2022,10,8,Wednesday,Saturday


### Remove duplicate rows using .drop_duplicates() or Excel’s “Remove Duplicates”.

In [27]:
# Check how many duplicate rows exist
duplicates = df_raw_dataset_Healthcare_Patient_Records.duplicated().sum()

print("Number of duplicate rows:", duplicates)

Number of duplicate rows: 100


In [28]:
# Remove duplicate rows
df_raw_dataset_Healthcare_Patient_Records = (
    df_raw_dataset_Healthcare_Patient_Records.drop_duplicates()
)

In [29]:
# check the duplicate again
print(
    "Duplicate rows after cleaning:",
    df_raw_dataset_Healthcare_Patient_Records.duplicated().sum()
)

Duplicate rows after cleaning: 0


In [30]:
# Check the number of rows before and after
print("Rows after removing duplicates:",
      len(df_raw_dataset_Healthcare_Patient_Records))

Rows after removing duplicates: 9900


In [31]:
# Check final shape
print("Final dataset shape:",
      df_raw_dataset_Healthcare_Patient_Records.shape)

Final dataset shape: (9900, 38)


### Identify and handle missing values using .isnull() in Python or filters in Excel

In [32]:
# Check total missing values in each column
missing_values = df_raw_dataset_Healthcare_Patient_Records.isnull().sum()

print(missing_values)

patient_id               0
patient_name             0
age                    197
gender                 192
blood_group            197
city                   195
department             199
diagnosis              200
admission_date           0
discharge_date           0
length_of_stay           0
heart_rate             200
systolic_bp              0
diastolic_bp             0
temperature_c            0
spo2                     0
respiratory_rate         0
glucose_mg_dl          197
cholesterol_mg_dl        0
hemoglobin_g_dl          0
bmi                    198
smoking_status          59
diabetes                 0
hypertension             0
medication            1610
treatment_cost_inr       0
insurance_type         199
doctor_id                0
outcome                  0
admission_type           0
admission_year           0
admission_month          0
admission_day            0
discharge_year           0
discharge_month          0
discharge_day            0
admission_day_name       0
d

**Interpretation**
- we have  column of that have medication its have 1610 missing null value

In [33]:
missing_percentage = (
    df_raw_dataset_Healthcare_Patient_Records.isnull().mean() * 100
).round(2)

print(missing_percentage[missing_percentage > 0])

age                1.99
gender             1.94
blood_group        1.99
city               1.97
department         2.01
diagnosis          2.02
heart_rate         2.02
glucose_mg_dl      1.99
bmi                2.00
smoking_status     0.60
medication        16.26
insurance_type     2.01
dtype: float64


In [34]:
# Handle numerical missing values
numeric_columns = [
    'age',
    'heart_rate',
    'glucose_mg_dl',
    'bmi'
]

for col in numeric_columns:
    df_raw_dataset_Healthcare_Patient_Records[col] = (
        df_raw_dataset_Healthcare_Patient_Records[col]
        .fillna(df_raw_dataset_Healthcare_Patient_Records[col].median())
    )

In [35]:
# Handle categorical missing values
categorical_columns = [
    'gender',
    'blood_group',
    'city',
    'department',
    'diagnosis',
    'smoking_status',
    'insurance_type'
]

for col in categorical_columns:
    df_raw_dataset_Healthcare_Patient_Records[col] = (
        df_raw_dataset_Healthcare_Patient_Records[col]
        .fillna(df_raw_dataset_Healthcare_Patient_Records[col].mode()[0])
    )

In [36]:
# check the result 
missing_after = df_raw_dataset_Healthcare_Patient_Records.isnull().sum()

print(missing_after[missing_after > 0])

medication    1610
dtype: int64


In [37]:
# for the  medication
df_raw_dataset_Healthcare_Patient_Records['medication'] = (
    df_raw_dataset_Healthcare_Patient_Records['medication']
    .fillna('unknown')
)

In [39]:
# check the result
missing_after = df_raw_dataset_Healthcare_Patient_Records.isnull().sum()

print(missing_after[missing_after > 0])


Series([], dtype: int64)


In [40]:
# Check total missing values in each column
missing_values = df_raw_dataset_Healthcare_Patient_Records.isnull().sum()

print(missing_values)

patient_id            0
patient_name          0
age                   0
gender                0
blood_group           0
city                  0
department            0
diagnosis             0
admission_date        0
discharge_date        0
length_of_stay        0
heart_rate            0
systolic_bp           0
diastolic_bp          0
temperature_c         0
spo2                  0
respiratory_rate      0
glucose_mg_dl         0
cholesterol_mg_dl     0
hemoglobin_g_dl       0
bmi                   0
smoking_status        0
diabetes              0
hypertension          0
medication            0
treatment_cost_inr    0
insurance_type        0
doctor_id             0
outcome               0
admission_type        0
admission_year        0
admission_month       0
admission_day         0
discharge_year        0
discharge_month       0
discharge_day         0
admission_day_name    0
discharge_day_name    0
dtype: int64


In [41]:
# show the first five file observations of given data frame 
df_raw_dataset_Healthcare_Patient_Records.head()

,patient_id,patient_name,age,gender,blood_group,city,department,diagnosis,admission_date,discharge_date,...,outcome,admission_type,admission_year,admission_month,admission_day,discharge_year,discharge_month,discharge_day,admission_day_name,discharge_day_name
0,P106252,Sneha Singh,31,male,ab+,ahmedabad,oncology,gastritis,2025-04-18,2025-04-20,...,recovered,elective,2025,4,18,2025,4,20,Friday,Sunday
1,P104684,Vivaan Reddy,38,male,b-,kolkata,orthopedics,arthritis,2022-05-09,2022-05-12,...,improved,emergency,2022,5,9,2022,5,12,Monday,Thursday
2,P101731,Isha Mehta,67,male,ab-,rajkot,neurology,diabetes,2024-09-12,2024-09-23,...,referred,emergency,2024,9,12,2024,9,23,Thursday,Monday
3,P104742,Rahul Koli,53,male,o+,ahmedabad,orthopedics,gastritis,2023-04-07,2023-04-13,...,referred,urgent,2023,4,7,2023,4,13,Friday,Thursday
4,P104521,Vikram Desai,39,male,b+,vadodara,oncology,diabetes,2022-10-05,2022-10-08,...,recovered,elective,2022,10,5,2022,10,8,Wednesday,Saturday
